azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -59,95 +64,152 @@ def __init__(self, wire_client, full_goal_state=False, base_incarnation=None):
+            raise
+
+    def _retrieve_certificates(self):
+        try:
+            uri = findtext(self._xml_doc, "Certificates")
+            if uri is not None:
+                xml_text = self._wire_client.fetch_config(uri, self._wire_client.get_header_for_cert())
                 self.certs = Certificates(xml_text)
+        except Exception as e:
+            self._log_document_retrieval_failure("certificates", e)
+            raise
 
-            uri = findtext(xml_doc, "ExtensionsConfig")
+    def _retrieve_shared_conf(self):
+        try:
+            uri = findtext(self._xml_doc, "SharedConfig")
             if uri is None:
-                self.ext_conf = ExtensionsConfig(None)
+                logger.warn("SharedConfig url doesn't exist in goal state")
             else:
-                xml_text = wire_client.fetch_config(uri, wire_client.get_header())
-                self.ext_conf = ExtensionsConfig(xml_text)
+                xml_text = self._wire_client.fetch_config(uri, self._wire_client.get_header())
+                self.shared_conf = SharedConfig(xml_text)
+        except Exception as e:
+            self._log_document_retrieval_failure("shared config", e)
+            raise
 
+    def _retrieve_remote_access(self):
+        try:
+            container = find(self._xml_doc, "Container")
             uri = findtext(container, "RemoteAccessInfo")
             if uri is None:
                 self.remote_access = None
             else:
-                xml_text = wire_client.fetch_config(uri, wire_client.get_header_for_cert())
+                xml_text = self._wire_client.fetch_config(uri, self._wire_client.get_header_for_cert())
                 self.remote_access = RemoteAccess(xml_text)
         except Exception as e:
-            logger.warn("Fetching the goal state failed: {0}", ustr(e))
+            self._log_document_retrieval_failure("remove access", e)
             raise
-        finally:
-            logger.info('Fetch goal state completed')
 
-    @staticmethod
-    def fetch_goal_state(wire_client):
-        """
-        Fetches the goal state, not including any nested properties (such as extension config).
-        """
-        return GoalState(wire_client)
+    def _retrieve_ext_conf(self):
+        try:
+            uri = findtext(self._xml_doc, "ExtensionsConfig")
+            self.ext_conf = self._ext_config_retriever.get_ext_config(self.incarnation, uri)
+        except Exception as e:
+            self._log_document_retrieval_failure("extensions config", e)
+            raise
 
-    @staticmethod
-    def fetch_full_goal_state(wire_client):
-        """
-        Fetches the full goal state, including nested properties (such as extension config).
-        """
-        return GoalState(wire_client, full_goal_state=True)
+    def _log_document_retrieval_failure(self, component_name, e):
+        estr = ustr(e)
+        if self._HttpFailedIndicator in estr:
```

Comment:
checking the exception message may not be robust enough, can you check on the exception type instead?

Reply:
Checking if it's a ProtocolError